### PR TITLE
[v2] Fixed work with flow >= 0.85, update to 0.91

### PR DESCRIPTION
**Pull request by DragorWW**
[Fixed work with flow >= 0.85](https://github.com/JedWatson/react-select/issues/3336), update to 0.90

[Since with flow 0.85, it has become stricter](https://medium.com/flow-type/asking-for-required-annotations-64d4f9c1edf8), and requires explicit typing of all export methods and variables.

### Changelog to support flow
After the upgrade, flow is displayed 58 new issues, most of them refer to external library and to the HOC manageState.

- add to flow ignore **@atlaskit/layer-manager** due to broken types *(components/FocusLock/index.js.flow, components/gateway/components/Gateway.js.flow, components/withRenderTarget.js.flow)*

Fixed typing in code and also made minor changes in code of docs

### Logic problems in code was fixed
New version flow found the following errors in logics:

- manageState - pass props to Select component incorrectly
- Select props isMulti:
  - There were problems due to the non-explicit work of the select
- Creatable: isValidNewOption
- Select: getOptionLabel and getOptionValue return React.Node

All changes are needed to work with the  flow >= 0.85

### Other moments
- SelectAnimated pass incorrect props to Select.componets

### P.S.
In my opinion, the required change should not have been so large, a lot of work was done because of old versions of libraries in dependencies

Idea of ​​using types from the npm library *(not from flow-typed)* is bad, and creates problems

Perhaps it makes sense to divide the pull request into 2 or more, for example, separately babel and separate flow

I am waiting for feedback, maybe I missed something, ready to discuss and correct

### CODE REVIEW

**Comment by DragorWW** [comment]
In the near future, I’ll see what the problem is with reduced coverage. Possible problem after updating babel or there is an error in the code.

**Comment by DragorWW** [comment]
I checked the problem with reduced coverage.

The reason is that `istanbul` with `babel@6` calculates it now incorrectly

**Example**: `src/components/Menu.js`
```js
49 [√] |export function getMenuPlacement(...) {
58 [√] |    const { spacing } = theme;
       |   // ...
63 [√] |    if (!menuEl || !menuEl.offsetParent) return defaultState;
67 [X] |    const { height: scrollHeight } = scrollParent.getBoundingClientRect();
       |    // ...
90 [X] |    switch (placement) {
       |        // ...
152    |        case 'top':
153    |        // 1: the menu will fit, do nothing
154[√] |        if (viewSpaceAbove >= menuHeight) {
155[√] |            return { placement: 'top', maxHeight };
156    |        }
       |     // ...
       |    }
       |    // ...
       |}
```

`Switch` *(line 154)* cannot be covered because the code after the first `return` *(line 67)* is not used

So now coverage is calculated correctly in the version of `babel@7`

**real coverage is 64%**

**Comment by gwyneplaine** [comment]
thanks for this @DragorWW, will have a more detailed look at this shortly. The write up and a cursory glance of the diff lgtm so far.

File: .babelrc [hunk 1]
```diff
@@ -1,9 +1,15 @@
 {
   "plugins": [
     "emotion",
-    "transform-class-properties",
-    "transform-object-rest-spread"
+    "@babel/plugin-proposal-class-properties",
+    "@babel/plugin-proposal-object-rest-spread"
```
Comment (TrySound) [inline .babelrc]: object-rest-spread is enabled in env preset by default

File: docs/App/Footer.js [hunk 2]
```diff
@@ -5,7 +5,7 @@ import React from 'react';
 // const smallDevice = '@media (max-width: 769px)';
 const largeDevice = '@media (min-width: 770px)';
 
-const Wrapper = props => (
+const Wrapper = (props: any) => (
```
Comment (TrySound) [inline docs/App/Footer.js]: Why? This is unsafe

File: rollup.config.js [hunk 3]
```diff
@@ -22,12 +22,16 @@ const external = Object.keys(globals);
...
+      '@babel/preset-flow',
+    ],
     plugins: [
       'emotion',
-      'transform-class-properties',
-      'transform-object-rest-spread',
-      'external-helpers',
+      '@babel/plugin-proposal-class-properties',
+      '@babel/plugin-proposal-object-rest-spread',
+      '@babel/plugin-external-helpers',
```
Comment (TrySound) [inline rollup.config.js]: @babel/plugin-external-helpers is not necessary
@babel/plugin-proposal-object-rest-spread is not necessary

File: src/Async.js [hunk 4]
```diff
@@ -16,6 +21,9 @@ export type AsyncProps = {
   /* If cacheOptions is truthy, then the loaded data will be cached. The cache
      will remain until `cacheOptions` changes value. */
   cacheOptions: any,
+  onInputChange: (string, InputActionMeta) => void,
+  filterOption?: null,
```
Comment (TrySound) [inline src/Async.js]: Does this field accept only null?

File: src/Async.js [hunk 5]
```diff
@@ -34,14 +42,16 @@ type State = {
   passEmptyOptions: boolean,
 };
 
-export const makeAsyncSelect = (SelectComponent: ComponentType<*>) =>
-  class Async extends Component<Props, State> {
+export const makeAsyncSelect = <C: {}>(
+  SelectComponent: AbstractComponent<C>
+): AbstractComponent<C & AsyncProps> =>
```
Comment (TrySound) [inline src/Async.js]: I would use spread instead

**Comment by TrySound** [comment]
I would better split build tools upgrade and flow types upgrade. Two PRs will be easier to review.

**Comment by DragorWW** [comment]
@TrySound yes it is a good idea, I will move build tools upgrade in a separate pr

**Comment by gwyneplaine** [comment]
@DragorWW, ty for this, can you rebase from master?

File: src/Async.js (hunk 5 above)
Comment (DragorWW) [reply src/Async.js]: spread is a good idea if you have exact flow types

**Comment by DragorWW** [comment]
> @DragorWW, ty for this, can you rebase from master?

rebased to master, and squash commits to one, and already removed some of the outdated edits and minor changes in the code style that were not related to the main task.

**Comment by DragorWW** [comment]
> @DragorWW, ty for this, can you rebase from master?

Rebase to master, and squash commits to one, and already removed some of the outdated edits and minor changes in the code style that were not related to the main task


What do you think about adding flow check to CI?

**Comment by TrySound** [comment]
sure

**Comment by gwyneplaine** [comment]
thanks @DragorWW @TrySound this looks good, I'm merging now 🎉

**Pull request merged**